File: services/runtime/runtime.go
// Package runtime provides the LLM gateway service for completion and embedding requests.
package runtime

// Message represents a chat message.
type Message struct {
	Role    string // system, user, assistant
	Content string
	Name    string // optional
}

// RoutingStrategy determines how to select providers.
type RoutingStrategy int

const (
	RoutingUnspecified RoutingStrategy = iota
	RoutingCostOptimized
	RoutingLatencyOptimized
	RoutingQualityOptimized
	RoutingSpecificProvider
)

// CompletionParams contains parameters for a completion request.
type CompletionParams struct {
	PromptRef    string            // e.g., "summarizer:v2.1"
	Messages     []Message
	Variables    map[string]string
	Routing      RoutingStrategy
	Provider     string // required if RoutingSpecificProvider
	Model        string
	Temperature  float64
	MaxTokens    int
	TopP         float64
	Stop         []string
	OutputSchema string // JSON schema for structured output
	UseCache     bool
	CacheTTL     int // seconds
	Metadata     map[string]string
}

// CompletionResult contains the result of a completion request.
type CompletionResult struct {
	ID               string
	Content          string
	Message          Message
	StructuredOutput map[string]interface{}
	Provider         string
	Model            string
	Usage            Usage
	Cached           bool
	TraceID          string
}

// Usage contains token usage information.
type Usage struct {
	PromptTokens     int
	CompletionTokens int
	TotalTokens      int
	CostUSD          float64
}

// StreamChunk represents a chunk of a streaming response.
type StreamChunk struct {
	ID       string
	Delta    string
	Done     bool
	Message  *Message // only on final chunk
	Usage    *Usage   // only on final chunk
	Provider string
	Model    string
	TraceID  string
	Err      error // non-nil if stream encountered an error
}

// EmbedParams contains parameters for an embedding request.
type EmbedParams struct {
	Texts    []string
	Model    string
	Provider string
}

// EmbedResult contains the result of an embedding request.
type EmbedResult struct {
	Embeddings []Embedding
	Model      string
	Provider   string
	Usage      Usage
}

// Embedding represents a single embedding vector.
type Embedding struct {
	Values     []float32
	Dimensions int
}

// ProviderInfo represents information about an LLM provider.
type ProviderInfo struct {
	Name            string
	Models          []string
	Available       bool
	CostPer1KTokens map[string]float64 // model -> cost
}

File: sdk/python/src/delos/models/runtime.py
"""Models for the runtime service."""

from __future__ import annotations

from enum import Enum

from pydantic import BaseModel, Field


class RoutingStrategy(str, Enum):
    """Strategy for routing requests to providers."""

    UNSPECIFIED = "unspecified"
    COST = "cost"
    LATENCY = "latency"
    QUALITY = "quality"


class Message(BaseModel):
    """A message in a conversation."""

    role: str  # system, user, assistant
    content: str


class CompletionParams(BaseModel):
    """Parameters for a completion request."""

    model: str = ""
    messages: list[Message] = []
    system_prompt: str = ""
    max_tokens: int = 1024
    temperature: float = 0.7
    top_p: float = 1.0
    stop_sequences: list[str] = []
    provider: str = ""  # optional: openai, anthropic
    routing_strategy: RoutingStrategy = RoutingStrategy.UNSPECIFIED
    metadata: dict[str, str] = {}


class Usage(BaseModel):
    """Token usage information."""

    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int = 0


class CompletionResponse(BaseModel):
    """Response from a completion request."""

    id: str = ""
    content: str = ""
    model: str = ""
    provider: str = ""
    usage: Usage = Field(default_factory=Usage)
    latency_ms: float = 0.0
    finish_reason: str = ""
    metadata: dict[str, str] = {}


class Model(BaseModel):
    """Information about an available model."""

    id: str
    name: str
    provider: str
    context_window: int = 0
    max_output_tokens: int = 0
    supports_vision: bool = False
    supports_function_calling: bool = False
    cost_per_input_token: float = 0.0
    cost_per_output_token: float = 0.0


class Provider(BaseModel):
    """Information about an LLM provider."""

    id: str
    name: str
    models: list[Model] = []
    is_available: bool = True
